### PR TITLE
feat(launch): allow users to specify job name via the `job_name` setting

### PR DESCRIPTION
**Pull request by bcsherma**
Fixes
-----
- Fixes WB-14253

Description
-----------

This PR introduces a new environment variable `WANDB_JOB_NAME` and `wandb.init` kwarg `job_name` that each allow users to specify the name of the job that will be created from a run.

### Summary

This pull request adds a new feature to wandb that allows users to specify a name for a job Artifact created from a wandb run. It updates the `wandb_settings.proto` file and the corresponding Python modules to support the new `job_name` field in the `Settings` message. It also adds logic to the `JobBuilder` class and the `wandb.init` function to handle the `job_name` parameter. It updates the documentation and the tests to reflect the new feature.

> _Sing, O Muse, of the glorious deeds of wandb,_
> _The mighty tool that tracks and logs the runs of code,_
> _How they added a new feature, `job_name`, to their settings,_
> _And gave the users power to name their artifacts of old._

### Walkthrough
*  Add a new feature to allow setting a job name in the wandb settings, which is used to create a job Artifact from a wandb run ([link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-68c7b8098da8be40d64502b236bfcb465b631d1f66f17b1ef789d752aeabedc7R719), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-f4264fdd3a281510ed93798673eea3c6b4456f413856c6c836758243dcb6e50bR270), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-6abcd261dffa10cad309f6faeb1699cbcbae670d0193a2435863642d8320c63bL18-R18), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-6abcd261dffa10cad309f6faeb1699cbcbae670d0193a2435863642d8320c63bL100-R100), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-94f50ee17a24613a3b82f34f4725782a6197b222b0c5db2f7daea0de595cd40bR203), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-94f50ee17a24613a3b82f34f4725782a6197b222b0c5db2f7daea0de595cd40bR430-R431), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-94f50ee17a24613a3b82f34f4725782a6197b222b0c5db2f7daea0de595cd40bR638), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-94f50ee17a24613a3b82f34f4725782a6197b222b0c5db2f7daea0de595cd40bL697-R702), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-7b68eb374de36a92776b31d45875dfe9a02f873c2f1a2fdf421c69fec8381a8bL17-R17), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-7b68eb374de36a92776b31d45875dfe9a02f873c2f1a2fdf421c69fec8381a8bL41-R41), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-dd44e141d4cf94de8e5470056ab14ca8cee6e5cd9986fab529929c5b358389fbR210), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-dd44e141d4cf94de8e5470056ab14ca8cee6e5cd9986fab529929c5b358389fbR437-R438), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-dd44e141d4cf94de8e5470056ab14ca8cee6e5cd9986fab529929c5b358389fbR645), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-dd44e141d4cf94de8e5470056ab14ca8cee6e5cd9986fab529929c5b358389fbL704-R709), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-aeac218eae0573cefa702ae803a36e0521c0dbcdd3c4344309009ba206e109b4R109), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-40f8cb407c2daff98227254c1110d2cad6313a355b82a00a989d3f40225ad19cR96), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-63f38ce7b0a29a84c0252e3ff04f02b546e4f199c0006956408db2c1fc567085R916), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-63f38ce7b0a29a84c0252e3ff04f02b546e4f199c0006956408db2c1fc567085R1014-R1016), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-432d500a1d77db4e4a22825d98e99cb860fee78777f6540b85395c1e5ee28e94R381), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-432d500a1d77db4e4a22825d98e99cb860fee78777f6540b85395c1e5ee28e94R738), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-432d500a1d77db4e4a22825d98e99cb860fee78777f6540b85395c1e5ee28e94R1548), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-432d500a1d77db4e4a22825d98e99cb860fee78777f6540b85395c1e5ee28e94R1735))
*  Update the `JobBuilder` class in `wandb/sdk/internal/job_builder.py` to use the job name from the settings if provided, or generate a default name based on the source of the job ([link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-16cc00c70034fdd17e6905e7821f5586abc8725c2904420cc5d95fa35f896bc7L187-R190), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-16cc00c70034fdd17e6905e7821f5586abc8725c2904420cc5d95fa35f896bc7L227-R233), [link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-16cc00c70034fdd17e6905e7821f5586abc8725c2904420cc5d95fa35f896bc7L243-R252))
*  Add a unit test for the job name feature in `tests/pytest_tests/unit_tests/test_library_public.py` ([link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-f4264fdd3a281510ed93798673eea3c6b4456f413856c6c836758243dcb6e50bR270))
*  Add a system test for the job name feature in `tests/pytest_tests/system_tests/conftest.py` ([link](https://github.com/wandb/wandb/pull/5791/files?diff=unified&w=0#diff-68c7b8098da8be40d64502b236bfcb465b631d1f66f17b1ef789d752aeabedc7R719))

### CODE REVIEW

**Review by KyleGoyette — COMMENTED** [review]
I think the arg should be job_name not just job

File: wandb/sdk/wandb_init.py [hunk 1]
```diff
@@ -1017,6 +1018,9 @@ def init(
             organize runs together into a larger whole, and you can toggle this
             on and off in the UI. For more details, see our
             [guide to grouping runs](https://docs.wandb.com/guides/runs/grouping).
+        job: (str, optional) Specify a name for a job Artifact created from this
```
[the remark this answers is not in the document]
Comment (bcsherma) [reply wandb/sdk/wandb_init.py]: Done. I feel like job is slick but it's not quite specific enough.

File: wandb/sdk/internal/job_builder.py [hunk 2]
```diff
@@ -167,7 +167,10 @@ def _build_repo_job(
             },
         }
 
-        name = make_artifact_name_safe(f"job-{remote}_{program_relpath}")
+        if hasattr(self._settings, "job_name") and self._settings.job_name:
```
Comment (gtarpenning) [inline wandb/sdk/internal/job_builder.py:170]: this is sort of a strange pattern, when do runs not have the attr at all?

**Review by gtarpenning — APPROVED** [review]
perfect, will work well with create job changes (I think).

File: wandb/sdk/internal/job_builder.py (hunk 2 above)
Comment (KyleGoyette) [reply wandb/sdk/internal/job_builder.py:170]: offline mode can cause that, if the static settings are not set right

Comment (bcsherma) [reply wandb/sdk/internal/job_builder.py:170]: the settings objects attributes are created dynamically such that if the user hasn't set the argument then the settings object here will not have the attribute. 🤷🏻‍♂️

File: wandb/proto/wandb_settings.proto [hunk 3]
```diff
@@ -106,6 +106,7 @@ message Settings {
   ListStringValue ignore_globs = 78;
   google.protobuf.DoubleValue init_timeout = 79;
   google.protobuf.BoolValue is_local = 80;
+  google.protobuf.StringValue job_name = 145;
```
Comment (bcsherma) [inline wandb/proto/wandb_settings.proto:109]: Not sure if this is the right way to add a value or if I should set it to 81 and increment the rest of the numbers.

File: wandb/sdk/internal/job_builder.py (hunk 2 above)
Comment (bcsherma) [reply wandb/sdk/internal/job_builder.py:170]: Looks like @dmitryduev did a settings refactor today so I'm not sure if this is still true.

File: wandb/proto/wandb_settings.proto (hunk 3 above)
Comment (KyleGoyette) [reply wandb/proto/wandb_settings.proto:109]: For settings I'm not sure, we should ask @dmitryduev who appears to have just added this?

File: wandb/sdk/internal/job_builder.py (hunk 2 above)
Comment (KyleGoyette) [reply wandb/sdk/internal/job_builder.py:187]: I think this is no longer necessary thanks to Dima's change

**Review by KyleGoyette — APPROVED** [review]
LGTM assuming @dmitryduev is agrees with how settings proto is modified (it is weird to have 145ish number next to 81)

File: wandb/proto/wandb_settings.proto (hunk 3 above)
Comment (dmitryduev) [reply wandb/proto/wandb_settings.proto:109]: the way you did it is great! we're not relying on field numbers anywhere internally

File: wandb/sdk/internal/job_builder.py (hunk 2 above)
Comment (dmitryduev) [reply wandb/sdk/internal/job_builder.py:187]: +1 to @KyleGoyette, `self._settings.job_name` should just work (it's a bug if it doesn't).

File: wandb/sdk/wandb_init.py [hunk 4]
```diff
@@ -913,6 +913,7 @@ def _attach(
 
 
 def init(
+    job_name: Optional[str] = None,
```
Comment (dmitryduev) [inline wandb/sdk/wandb_init.py:916]: Why do you need it as a new arg in init? We're trying not to be adding these unless absolutely necessary. We've been putting such stuff to Settings, so `wandb.init(settings=wandb.Settings(job_name=...))` would have the same effect. One caveat is documenting the Settings object, but we're getting there :)

**Review by dmitryduev — COMMENTED** [review]
Thanks @bcsherma. I'd prefer not adding a new init arg, but maybe I'm missing something -- please LMK.

File: wandb/sdk/wandb_init.py (hunk 4 above)
Comment (bcsherma) [reply wandb/sdk/wandb_init.py:916]: Adding that kwarg is the main purpose of this PR. To make it easier for users to set a name for their job when it is created. We can just put it in settings but I don't think most users know that the settings object even exists. I think a wandb.init kwarg is far more discoverable and clear.

Comment (bcsherma) [reply wandb/sdk/wandb_init.py:916]: @dmitryduev can you help me understand why you prefer not to introduce new `wandb.init` kwargs unless absolutely necessary?

Comment (bcsherma) [reply wandb/sdk/wandb_init.py]: We are removing the argument. Job name will be configurable by settings and WANDB_JOB_NAME until we revamp the API

**Review by gtarpenning — APPROVED** [review]
nice

File: wandb/sdk/internal/job_builder.py [hunk 5]
```diff
@@ -207,8 +211,6 @@ def _build_repo_job_source(
                     "notebook": metadata["notebook"],
                 }
             )
-        name = make_artifact_name_safe(f"job-{remote}_{program_relpath}")
-
```
[the remark this answers is not in the document]
Comment (bcsherma) [reply wandb/sdk/internal/job_builder.py:211]: no the order doesn't really matter

File: wandb/sdk/wandb_settings.py [hunk 6]
```diff
@@ -1543,6 +1545,7 @@ def _apply_env_vars(
             "WANDB_NOTES": "run_notes",
             "WANDB_TAGS": "run_tags",
             "WANDB_JOB_TYPE": "run_job_type",
+            "WANDB_LAUNCH_JOB_NAME": "job_name",
```
Comment (dmitryduev) [inline wandb/sdk/wandb_settings.py:1548]: Why not call the setting `launch_job_name`? No such mapping would be required then (any env var starting with WANDB_ gets mapped to a setting automatically).

File: wandb/sdk/wandb_settings.py [hunk 7]
```diff
@@ -1729,6 +1731,7 @@ def _apply_init(self, init_settings: Dict[str, Union[str, int, None]]) -> None:
             job_type="run_job_type",
             notes="run_notes",
             dir="root_dir",
+            job_name="job_name",
```
Comment (dmitryduev) [inline wandb/sdk/wandb_settings.py:1734]:
```suggestion
```

**Review by dmitryduev — APPROVED** [review]

File: wandb/sdk/wandb_settings.py (hunk 7 above)
Comment (bcsherma) [reply wandb/sdk/wandb_settings.py]: Good catch!

**Pull request merged**